controller/src/main/java/io/pravega/controller/store/stream/ZkOrderedStore.java

```diff
@@ -0,0 +1,317 @@
+                    } else {
+                        log.debug("entity {} added for stream {}/{} at position {}", entity, scope, stream, r);
+                    }
+                });
+    }
+
+    /**
+     * Method to remove entities from the ordered set. Entities are referred to by their position pointer.  
+     * @param scope scope
+     * @param stream stream
+     * @param entities list of entities' positions to remove
+     * @return CompletableFuture which when completed will indicate that entities are removed from the set. 
+     */
+    CompletableFuture<Void> removeEntities(String scope, String stream, Collection<Long> entities) {
+        Set<Integer> collections = entities.stream().collect(Collectors.groupingBy(x -> new Position(x).collectionNumber)).keySet();
+        return Futures.allOf(entities.stream()
+                                     .map(entity -> storeHelper.deletePath(getEntityPath(scope, stream, entity), false))
+                                     .collect(Collectors.toList()))
+                      .thenCompose(v -> Futures.allOf(
+                              collections.stream().map(collectionNum -> isSealed(scope, stream, collectionNum)
+                                      .thenCompose(sealed -> {
+                                          if (sealed) {
+                                              return tryDeleteSealedCollection(scope, stream, collectionNum);
+                                          } else {
+                                              return CompletableFuture.completedFuture(null);
+                                          }
+                                      })).collect(Collectors.toList())))
+                      .whenComplete((r, e) -> {
+                          if (e != null) {
+                              log.error("error encountered while trying to remove entity positions {} for stream {}/{}", entities, scope, stream, e);
+                          } else {
+                              log.debug("entities at positions {} removed for stream {}/{}", entities, scope, stream);
+                          }
+                      });
+    }
+
+    /**
+     * Returns a map of position to entity that was added to the set. 
+     * Note: Entities are ordered by position in the set but the map responded from this api is not ordered by default.
+     * Users can filter and order elements based on the position and entity id.  
+     * @param scope scope scope
+     * @param stream stream stream
+     * @return CompletableFuture which when completed will contain all positions to entities in the set. 
+     */
+    CompletableFuture<Map<Long, String>> getEntitiesWithPosition(String scope, String stream) {
+        Map<Long, String> result = new ConcurrentHashMap<>();
+        return Futures.exceptionallyExpecting(storeHelper.getChildren(getStreamPath(scope, stream)), DATA_NOT_FOUND_PREDICATE, Collections.emptyList())
```

Comment:
One clarification here. We are getting the children of a given stream:

```
storeHelper.getChildren(getStreamPath(scope, stream)
```

and iterating over the children assuming they are all collections, is it really the case, isn't there any other children which are not (parents of) collections?

Reply:
No, this is getting all children for a "stream" specific znode where all transactions that have been marked for "committing" are stored. 

The hierarchy is as follows:
```
     * This class creates a collection of collections starting with collection numbered 0.
     * A new collection is created under the znode: 
     * \<orderer-root\>/scope/stream/\<collectionNum\>/entities
     * It adds entries to lowest numbered collection using a persistent sequential znode 
     * until its capacity is exhausted. 
     * Once a collection is exhausted, a successor collection with number `current + 1` is created and current successor 
     * is marked as sealed by created following znode: 
     *     `\<orderer-root\>/scope/stream/\<collectionNum\>/sealed
```

So it gets all children of `\<orderer-root\>/scope/stream/` and recursively on 
`entities`. It uses collectionNum+entityNum to create a 64 bit "position" for each transaction.

Reply:
Oh I see, you are creating the collections directly under the `scope/stream/` znode. For organization purposes or simply to avoid having the creation of collections affecting the `cversion` of `scope/stream/`, you may want to have the collections under `scope/stream/collections`, in which case the entities path would be: `scope/stream/collections/<collectionNum>`. What do you think?

Reply:
done